### PR TITLE
Allow creating broadcast from nowhere

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Make it possible to create a new broadcast during scheduling. This new broadcast neither comes from a `BroadcastOp` nor from a fusion input. It just appears from no where. This PR only adds the minimal support to our system to make the simple pointwise fusion at test `BroadcastFromNowhere` work. So far, we can not inline such fusions. Our sync analysis could not handle this either. I am not working on these two pieces so far, because it is not clear to me whether this will be needed for scheduling hopper MMA.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/utils.cpp [hunk 1]
```diff
@@ -755,6 +755,14 @@ std::vector<IterDomain*> allIDsOf(const TensorView* tv) {
       auto all_vals_01 = DependencyCheck::getAllValsBetween(
           {dom0->begin(), dom0->end()}, {dom1->begin(), dom1->end()});
       all_vals.pushBack(all_vals_01);
+      // Getting all vals is not sufficient, because there might be broadcasting
```
Comment (naoyam) [inline csrc/ir/utils.cpp:758]: Does this capture new broadcast IDs if they are just inserted but not used?

Comment (zasdfgbnm) [reply csrc/ir/utils.cpp:758]: Good question, let me more thoroughly test it.

File: csrc/ir/internal_base_nodes.h [hunk 2]
```diff
@@ -683,6 +693,7 @@ class TensorDomain : public Val {
   const std::vector<IterDomain*> logical_domain_;
   std::vector<IterDomain*> allocation_domain_;
   std::vector<IterDomain*> loop_domain_;
+  std::vector<IterDomain*> additional_ids_;
```
Comment (zasdfgbnm) [inline csrc/ir/internal_base_nodes.h:696]: I believe this mechanism can be used for fixing:
https://github.com/NVIDIA/Fuser/pull/2379#discussion_r1653381767
But I will leave it as a future work.

File: csrc/ir/utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/ir/utils.cpp:763]: Added a new test. I do find some issues, so I added a new field `TensorDomain::additional_ids_` to help keep track of all IDs.

File: csrc/bfs.h [hunk 3]
```diff
@@ -145,7 +145,7 @@ class BFS {
 
     while (!allToNodesVisited()) {
       bool something_was_processed = false;
-      std::deque<NodeType> not_ready_;
+      std::deque<NodeType> not_ready;
```
Comment (zasdfgbnm) [inline csrc/bfs.h:148]: Just a renaming, no logical change

**Comment by zasdfgbnm** [comment]
@naoyam This is ready for review again

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/interface_nodes.h [hunk 4]
```diff
@@ -282,6 +282,9 @@ class NVF_API TensorView : public Val {
       int64_t position,
       ComputeAtMode mode = ComputeAtMode::Standard);
 
+  //! Create a new broadcast IterDoimain with extent one in the loop domain
```
Comment (naoyam) [inline csrc/ir/interface_nodes.h:285]: Let's say this is a 0-dim tensor. Would inserting a broadcast result in any semantic change? 

Asserting non 0-dim would be also reasonable.

Comment (zasdfgbnm) [reply csrc/ir/interface_nodes.h:285]: I think the "semantic change" is "we want to schedule it like a 1D broadcasting tensor"? Allowing 0-dim is aligned with PyTorch's `unsqueeze`:

```python
>>> torch.zeros(()).unsqueeze(0).shape
torch.Size([1])
>>> torch.zeros(()).unsqueeze(-1).shape
torch.Size([1])
```

File: csrc/ir/utils.h [hunk 5]
```diff
@@ -499,6 +499,10 @@ std::vector<TensorView*> getTVsWithDynamicTransform(Fusion* fusion);
 //! Then [I0, I1, I8, I9] is equivalent to [I6, I7, I2, I3]. [I0, I1, I2, I3] is
 //! equivalent to [I6, I7, I8, I9]. But [I0, I1, I8, I3] is NOT equivalent to
 //! [I6, I7, I2, I9]
+//!
+//! Broadcast IterDomains are ignored in this check, because we consider them as
+//! placeholders and allow them to be created (and annihilated?) arbitrarily as
```
Comment (zasdfgbnm) [inline csrc/ir/utils.h:504]: > and annihilated

Looks like the capability of creation and annihilation comes together, especially in the case today that we no longer assume the direction of transformations.

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/internal_base_nodes.h [hunk 6]
```diff
@@ -619,6 +626,9 @@ class TensorDomain : public Val {
   //! Returns a position of a root domain
   int64_t rootPosOf(IterDomain* id) const;
 
+  //! Create a new broadcast IterDoimain with extent one in the loop domain
```
Comment (jacobhinkle) [inline csrc/ir/internal_base_nodes.h:629]:
```suggestion
  //! Create a new broadcast IterDomain with extent one in the loop domain
```

File: csrc/ir/interface_nodes.h (hunk 4 above)
Comment (jacobhinkle) [inline csrc/ir/interface_nodes.h:285]:
```suggestion
  //! Create a new broadcast IterDomain with extent one in the loop domain
```

**Review by jacobhinkle — APPROVED** [review]
